related improvements to grm, rrm, hwe_normalized_pca, pca, toIRM

jbloom22 opened the pull request:
Builds on hwe changes in #3261 so also fixes #3239. These three methods now:
- take call expressions
- no longer have biallelic and variant requirements
- compute mean and variance from other stats per row, not per entry
- do not collect sample IDs on the Python side (grm, rrm).

GRM and RRM do still have the string col key requirement, but most of KinshipMatrix functionality still requires that. Plan for later is to kill KinshipMatrix altogether, with these methods producing BlockMatrices, and having ways to output files from BlockMatrix plus a table of string ids.

tpoterba commented on python/hail/methods/statgen.py:
```diff
@@ -2158,64 +2156,65 @@ def genetic_relatedness_matrix(dataset) -> KinshipMatrix:
-    dataset = dataset.annotate_rows(AC=agg.sum(dataset.GT.n_alt_alleles()),
-                                    n_called=agg.count_where(hl.is_defined(dataset.GT)))
-    dataset = dataset.filter_rows((dataset.AC > 0) & (dataset.AC < 2 * dataset.n_called)).persist()
+    mt = call_expr._indices.source
```
now's the time to abstract this chunk into a nice function somewhere.

tpoterba commented on src/main/scala/is/hail/methods/KinshipMatrix.scala:
```diff
@@ -2,29 +2,30 @@ package is.hail.methods
+    require(colKeys.signature.size == 1 && colKeys.signature.fields(0).typ.isOfType(TString()))
+    val sampleIds: Array[Annotation] = colKeys.collect().map(_.getAs[Annotation](0))
+    
+    KinshipMatrix(hc, TString(), matrix.toIndexedRowMatrix(), sampleIds, numVariantsUsed.longValue())
```
why does this still have a string requirement?

tpoterba replied:
can't it be the struct now?

jbloom22 replied:
See comment. This class is going away.

jbloom22 commented:
@tpoterba do I need to incorporate `base, _ = dataset._process_joins(entry_expr)` in `analyze_entry_expr`?

tpoterba commented on python/hail/methods/statgen.py:
```diff
@@ -1454,16 +1452,10 @@ def pca(entry_expr, k=10, compute_loadings=False, as_array=False) -> Tuple[List[
-    scores = Table(Env.hail().methods.PCA.scoresTable(dataset._jvds, as_array, r._2()))
+    mt = analyze_entry_expr('pca/entry_expr', entry_expr)
+
+    r = Env.hail().methods.PCA.apply(mt._jvds, to_expr(entry_expr)._ast.to_hql(), k, compute_loadings, as_array)
```
process joins is definitely required here. The solution would be to rework PCA to take a field name.

tpoterba replied:
you didn't address this, I think

tpoterba replied:
and can you add a test of something that uses a `literal`?

like `annotate-entries(x = x * hl.literal([1])[0])`

tpoterba commented:
oops, didn't realize I was actually assigned. will do a fuller review.

tpoterba commented on python/hail/tests/test_methods.py:
```diff
@@ -388,17 +388,29 @@ def hail_calculation(ds):
-        dataset = hl.balding_nichols_model(3, 100, 100)
-        eigenvalues, scores, loadings = hl.pca(dataset.GT.n_alt_alleles(), k=2, compute_loadings=True)
+        mt = hl.balding_nichols_model(3, 100, 50)
+        eigenvalues, scores, loadings = hl.pca(mt.GT.n_alt_alleles(), k=2, compute_loadings=True)
```
add the literal thing here to test the join

tpoterba replied:
still need to address this

jbloom22 commented:
I changed PCA and toIndexRowMatrix to take a field. Now these all use select_entries so no need to analyze keys or process joins. But I still check that the expression has a matrix table source with good error message using `matrix_table_source`.

jbloom22 commented:
fixed and rebased

jbloom22 commented:
Updated so PCA avoids select when the expression is a field (similar to BlockMatrix.write_from_entry_expr). Hopefully good to go now.

tpoterba commented on python/hail/linalg/blockmatrix.py:
```diff
@@ -387,21 +386,15 @@ def write_from_entry_expr(entry_expr, path, block_size=None):
-
-        if entry_expr in source._fields_inverse:
-            source._jvds.writeBlockMatrix(path, source._fields_inverse[entry_expr], block_size)
+        if entry_expr in mt._fields_inverse:
```
these two branches have different semantics. A row or column field will work in the 'else', because you can do `mt.select_entries(x = <row field>)`. I'm not sure which is correct (being strict or permissive) but it seems like this performance hack shouldn't have different behavior.

tpoterba replied:
mostly I don't really like the fields_inverse trick, though understand why it's done for now. I guess if this is marked with a fixme to delete ASAP, that's fine.

jbloom22 replied:
Doesn't this check in `select_entries` prevent a row field from working?
```
            if not e._indices == self._entry_indices:
                # detect row or col fields here
                raise ExpressionException("method 'select_entries' parameter 'exprs' expects entry-indexed fields,"
                                          " found indices {}".format(list(e._indices.axes)))
```

tpoterba replied:
only as varargs, not with kwargs.

tpoterba commented on python/hail/methods/statgen.py:
```diff
@@ -2158,64 +2147,61 @@ def genetic_relatedness_matrix(dataset) -> KinshipMatrix:
-    bm = BlockMatrix.from_entry_expr(normalized_genotype_expr)
-    dataset.unpersist()
+    bm = BlockMatrix.from_entry_expr(normalized_gt)
+    mt.unpersist()
```
I don't see a persist  call in this method. Should this be removed?

jbloom22 replied:
It's there in GRM and RRM, but I'll move persist to it's own line so easier to spot.

tpoterba commented on python/hail/methods/statgen.py:
```diff
@@ -1454,16 +1451,17 @@ def pca(entry_expr, k=10, compute_loadings=False, as_array=False) -> Tuple[List[
-    scores = Table(Env.hail().methods.PCA.scoresTable(dataset._jvds, as_array, r._2()))
+    mt = matrix_table_source('pca/entry_expr', entry_expr)
+
+    if entry_expr in mt._fields_inverse:
```
I don't see a super compelling performance reason for this. Compared to the SVD itself, an extra copy should be basically free. Not worth the added complexity (and semantics change, see above) IMO.

jbloom22 replied:
Isn't the concern more about needlessly invoking the compiler?

tpoterba replied:
at some point soon a `select_entries(__x = mt.x)` will be pretty much free.

tpoterba replied:
and if it's free, there's no reason to introduce extra complexity in the python

jbloom22 commented:
I've moved `check_entry_indexed` outside the conditional and added a note to remove the conditional entirely once select_entries on a field is free. I think it's reasonable to keep the conditional until then, at least in the block matrix case where needlessly invoking the compiler on every double may effect performance. If you feel strongly, I'll remove it from PCA and test the performance impact on write BlockMatrix (though I'd rather just leave the latter unperturbed).

tpoterba commented on python/hail/tests/test_methods.py:
```diff
@@ -388,18 +388,41 @@ def hail_calculation(ds):
+        eigenvalues, scores, loadings = hl.pca(mt.GT.n_alt_alleles(), k=1, compute_loadings=True)
+
+        mt = mt.annotate_entries(x = mt.GT.n_alt_alleles() * hl.literal([1])[0])
+        eigenvalues2 , scores2, loadings2 = hl.pca(mt.x, k=1, compute_loadings=True)
```
this doesn't test joined expressions coming into pca...

tpoterba replied:
you need to do it as one line, not break it up with an annotate_entries
